File: 1253-sort-the-matrix-diagonally/1253-sort-the-matrix-diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int rows = mat.size(); 
        int cols = mat[0].size();
        int noOfDiag = rows+cols-1; 
        int col = 0;
        int row = 0; 

        for(int d=1; d<=noOfDiag; d++){
            // initializing empty vector for storing diagonal element
            vector<int> diag;

            if(col == cols | row > 0){
                col = 0;
                row++; 
            }

            for(int i=row, j=col; i<rows && j<cols; i++, j++){
                diag.push_back(mat[i][j]);
            }
            
            sort(diag.begin(), diag.end());

            for(int i=row, j=col, k=0; i<rows && j<cols; i++, j++, k++){
                mat[i][j] = diag[k];
            }
            
            if(row == 0){
                col++;
            }

        }
        return mat;
    }
};

File: 1253-sort-the-matrix-diagonally/1253-sort-the-matrix-diagonally.py
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        rows = len(mat)
        cols = len(mat[0])
        noOfDiag = rows+cols-1
        row = 0
        col = 0

        for d in range(noOfDiag):
            diag = []

            if col == cols or row > 0:
                col = 0
                row += 1
            
            # storing diagonal elemnts into diag array
            i=row
            j=col
            while i<rows and j<cols:
                diag.append(mat[i][j])
                i += 1
                j += 1
            
            # sorting the diag array
            diag.sort()
            
            # replace the matrix diagonal elements with diag array elements
            i=row
            j=col
            k = 0
            while i<rows and j<cols:
                mat[i][j] = diag[k]
                i += 1
                j += 1
                k += 1
            
            if row == 0:
                col += 1
        
        return mat
